List CxJS Hacker News

mstijak opened the pull request:
Here's my entry based on CxJS. 

I had problems getting Lighthouse scores while using Firebase API (reported in #82 ), so I switched to Firebase REST API which resolved the issue. 

Performance score and timings are not perfect since the application is making 30 requests on load to fetch stories, but I decided to submit the entry anyway...

addyosmani commented:
@housseindjirdeh @anubhav7495 would either of you be up for auditing this implementation?

One issue I noticed was when going offline the AppShell appears to be correctly cached but entries are not being dynamically cached so you're unable to view them without a network connection.

housseindjirdeh commented:
@mstijak this is a solid implementation, good job mate :+1: 

Just a few things that may help:
1. As @addyosmani mentioned, the AppShell seems cached correctly and works offline but the endpoints aren't cached dynamically (if you try loading the app offline you'll notice that the requests fail). It seems you're using `SwPrecacheWebpackPlugin` and since you should be able to pass options from `sw-precache`, you should be able to use [`runtimeCaching`](https://github.com/GoogleChrome/sw-precache#runtimecaching-arrayobject) for this
2. Taking a look at [WPT runs](https://www.webpagetest.org/video/compare.php?tests=170723_JT_a9814305726a587ab49be67060c64146-r:1-c:0) and first paint times look promising. However just like you mentioned, there's room for improvement in your Lighthouse performance scores ([interactive times >8s on EM](https://www.webpagetest.org/lighthouse.php?test=170723_JT_a9814305726a587ab49be67060c64146&run=3)). Since you're using regular REST  requests for your API and you're not relying on an SDK with a WebSocket, you can try using another [unofficial API](https://github.com/cheeaun/node-hnapi) which will help reduce the number of requests made on each page significantly.

Just a couple of suggestions, otherwise the application looks good and I look forward to having it in HNPWA once a few things are smoothed out :raised_hands:

mstijak commented:
Thank you for looking into this. 
I followed your advice and switched to unofficial API. The results are much better now as you can see in the updated entry. The app is interactive in 4.5s for EM and 3.7s for 3G. The only minor annoyance is related to a small, but noticeable,  delay when loading stories with hundreds of comments. That worked better previously.
Adding `runtimeCaching` was very straightforward too. I didn't know that's possible. Thanks for the tip.
I also added the `/user` route and some basic error handling. 

I'm very happy with the improvements and in case you have more pro tips to share, I would love to play further :)

housseindjirdeh commented on site/_apps/cxjs-hackernews.txt:
```diff
@@ -0,0 +1,27 @@
+api: 'Node-hnapi (unoffical)'
+hosting: 'Netlify'
+authors: 
+    - name: 'Marko Stijak'
```
The name here will need to match the name in your `_contributors` submission in order to show you avatar underneath your app :)

housseindjirdeh commented on site/_apps/cxjs-hackernews.txt:
```diff
@@ -0,0 +1,27 @@
+---
+weight: 15
```
Could you please bump this number up to the weight of the most recent contribution  + 1?

housseindjirdeh commented on site/_contributors/marko.txt:
```diff
@@ -0,0 +1,6 @@
+---
+weight: 17
```
Similarly here (most recent contributor weight + 1)

mstijak commented:
@housseindjirdeh I fixed the files. Thanks for checking this out once again.

addyosmani commented:
@housseindjirdeh Did you have any remaining feedback on this one? If it looks good to go, happy to do a final review on perf numbers and land 👋

housseindjirdeh commented:
This looks good to go @addyosmani :)

Thanks again @mstijak 🙌

mstijak commented:
Thank you guys. I really appreciate your help. 

https://www.webpagetest.org/result/170823_S3_ac3c9bf1eb9f889c60ca20964061c468/
